Tutorial on decompiling objects (and also object_dns documented) (#647)
zeldaret/mm: docs/tutorial/object_decomp_example.md

# Object Decompilation Example

Previous: [Object Decompilation](object_decomp.md)

Let's take a look at `object_dns`, which is a pretty typical NPC object. It's used by one actor: `ovl_En_Dns`.

## Step 1: Naming the skeleton and limbs

> We already went through the steps of opening an object file in Z64Utils in the [documenting step](documenting.md#z64utils), but we'll do it one more time here. First, search for the object file, then right-click and select "Open in Object Analyzer". It will ask you to choose a segment: this is the segment that the file is put on, and allows Z64Utils to resolve the segmented addresses it references into symbols. The json already knows it should be segment 6, so just click OK.

Reviewer: ```suggestion
We already went through the steps of opening an object file in Z64Utils in the [documenting step](documenting.md#z64utils), but we'll do it in more detail here. First, search for the object file, then right-click and select "Open in Object Analyzer". It will ask you to choose a segment: this is the segment that the file is put on, and allows Z64Utils to resolve the segmented addresses it references into symbols. The json already knows it should be segment 6, so just click OK.
```

![Opening object_dns in Z64Utils](images/z64utils_open_dns.png)

In the resulting window, go to "Analysis -> Find Dlists" and press OK (the defaults are usually fine). This will only find display lists, so to find everything else in the obect, do "Analysis -> Analyze Dlists". We're looking for the skeleton, so scroll down to the SkeletonHeader, right-click it, and select "Open in Skeleton Viewer":

![Finding object_dns's SkeletonHeader in Z64Utils](images/z64utils_dns_skeletonheader.png)

> When you open the Skeleton Viewer, you'll see a list of animations off to the side. Selecting one of them will display an error that says something like "RENDER ERROR AT 0x06001A98! (Could not read 0x80 bytes at address 08000000)". This is because one of the display lists in the skeleton is expecting something to be set at segment 8. From the actor, we know that it's expecting the eye textures to be loaded into segment 8 like so:

Reviewer: ```suggestion
When you open the Skeleton Viewer, you'll see a list of animations off to the side. Selecting one of them will display an error that says something like `RENDER ERROR AT 0x06001A98! (Could not read 0x80 bytes at address 08000000)`. This is because one of the display lists in the skeleton is expecting something to be set at segment 8. From the actor, we know that it's expecting the eye textures to be loaded into segment 8 like so:
```

```c
static TexturePtr D_8092DE1C[] = { &D_060028E8, &D_06002968, &D_060029E8, &D_06002968 };
[...]
gSPSegment(POLY_OPA_DISP++, 0x08, Lib_SegmentedToVirtual(sEyeTextures[this->eyeIndex]));
```

Let's set segment 8 to be one of the eye textures listed here. Click on "Segments", then for segment 08, click "Edit" and select "Address" as the source. Input "060028E8" as the address, then hit OK. This should result in the window looking as follows:

![Setting segment 8 to one of the eye textures in Z64Utils](images/z64utils_set_segment_8.png)

> Now that we've gotten around the error, we can see what each limb in the skeleton corresponds to by clicking on it in the Hierarchy. Clicking on any given limb will highlight what part of the model it represents in red. Note that some limbs don't actually render anything, so sometimes clicking on a limb will not turn anything red. In the below example, this limb is clearly the head:

Reviewer 2: When you mention limbs that don't actually render, could you perhaps explain what to do in that case?

```

Note that this step might be tricky to do if multiple things in the actor use the same segment. It's okay to wait to do this until you've named all the display lists in the actor, since that will make it easier to find the display list associated with a given texture.

## Step #4: Naming anything else in the actor

For some actors, there may be a few other things left to name that are directly referenced in the actor's code. In our case, there is one display list that we still need to name:
```c
gSPDisplayList(POLY_OPA_DISP++, &D_06002C48);
```

In Z64Utils, scroll to find this display list, then right-click and select "Open in Dlist Viewer":

![Opening dlist_00002C48 in Z64Utils](images/z64utils_dns_display_list.png)

We can see this is the guard's Deku Flower:

![Showing the guard's Deku Flower in Z64Utils](images/z64utils_dns_deku_flower.png)

We can name the display list as such in the XML:
```xml
<DList Name="gKingsChamberDekuGuardDekuFlower" Offset="0x2C48" />
```

Then, like all steps before, we can run `./extract_assets.py -s objects/object_dns` and then update `z_en_dns.c` with our new name:
```c
gSPDisplayList(POLY_OPA_DISP++, gKingsChamberDekuGuardDekuFlower);
```

> Since we've now named every single thing in the actor, we can delete all symbols associated with `ovl_En_Dns` in [undefined_syms](../../undefined_syms.txt).

Reviewer 2: This shouldn't be necessary as it should already be done by actor decomp.

<DList Name="gKingsChamberDekuGuardLeftLeafDL" Offset="0x1CB0" />
<DList Name="gKingsChamberDekuGuardCenterLeafDL" Offset="0x1D40" />
<DList Name="gKingsChamberDekuGuardRightLeafDL" Offset="0x1DD8" />
```

Run `./extract_assets.py -s objects/object_dns` once again, since it will help in the next step to have all of our display lists named.

### Step #6: Naming remaining textures

With every display list named, it's now a lot easier to name the remaining textures. In the `assets/objects/object_dns/` folder, you can see all the textures in the object as various PNG files. For some of the textures, just looking at them will give you a good idea as to what they should be named. For other textures, it may help to see how the texture is used in the object's display lists. Let's take a look at `object_dns_Tex_002868`, which is only used in one display list:
```c
Gfx gKingsChamberDekuGuardSnoutDL[] = {
    [...]
    gsDPLoadTextureBlock(object_dns_Tex_002868, G_IM_FMT_RGBA, G_IM_SIZ_16b, 8, 8, 0, G_TX_MIRROR | G_TX_CLAMP,
                         G_TX_MIRROR | G_TX_CLAMP, 3, 3, G_TX_NOLOD, G_TX_NOLOD),
    [...]
};
```

> Since `tex_002868.rgba16.png` looks like the inside of the Deku Guard's mouth, and since it's *only* used in the snout display list, you can probably name this something like `gKingsChamberDekuGuardMouthTex`. But rather than guessing, we can confirm it by editing the texture ourselves and then viewing it in-game. Let's change `tex_002868.rgba16.png` to this:

Reviewer 2: ```suggestion
Since `tex_002868.rgba16.png` looks like the inside of the Deku Guard's mouth, and it's *only* used in the snout display list, you can probably name this something like `gKingsChamberDekuGuardMouthTex`. But rather than guessing, we can confirm it by editing the texture ourselves and then viewing it in-game. Let's change `tex_002868.rgba16.png` to this:
```
The second since seems redundant.

![Our custom mouth texture](images/dns_custom_texture.png)

> 

Reviewer 2: I think it would be beneficial show/explain exactly how you changed the texture. In particular the first bullet point of the investigative modding seems unclear to actually how to do it.

Author: I'm probably the worst person to ask about this sort of thing because I use Paint like an idiot, lol. But I took a shot at explaining it in both the investigative modding section and in here.
